vsp.go

```diff
@@ -60,62 +66,298 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
+
+	var nextIntervalStart, expiry int32
+	var cancels []func()
+	for {
+		select {
+		case <-ctx.Done():
+			defer outerCancel()
+			fatalMu.Lock()
+			err := fatal
+			fatalMu.Unlock()
+			if err != nil {
+				return err
+			}
+			return ctx.Err()
+		case n := <-c.C:
+			if len(n.AttachedBlocks) == 0 {
+				continue
+			}
+
+			tip := n.AttachedBlocks[len(n.AttachedBlocks)-1]
+			w := vsp.w.Internal()
+
+			// Don't perform any actions while transactions are not synced through
+			// the tip block.
+			rp, err := w.RescanPoint(ctx)
+			if err != nil {
+				return err
+			}
+			if rp != nil {
+				log.Debugf("Skipping autobuyer actions: transactions are not synced")
+				continue
+			}
+
+			tipHeader, err := w.BlockHeader(ctx, tip)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			height := int32(tipHeader.Height)
+
+			// Cancel any ongoing ticket purchases which are buying
+			// at an old ticket price or are no longer able to
+			// create mined tickets the window.
+			if height+2 >= nextIntervalStart {
+				for i, cancel := range cancels {
+					cancel()
+					cancels[i] = nil
+				}
+				cancels = cancels[:0]
+
+				intervalSize := int32(w.ChainParams().StakeDiffWindowSize)
+				currentInterval := height / intervalSize
+				nextIntervalStart = (currentInterval + 1) * intervalSize
+
+				// Skip this purchase when no more tickets may be purchased in the interval and
+				// the next sdiff is unknown.  The earliest any ticket may be mined is two
+				// blocks from now, with the next block containing the split transaction
+				// that the ticket purchase spends.
+				if height+2 == nextIntervalStart {
+					log.Debugf("Skipping purchase: next sdiff interval starts soon")
+					continue
+				}
+				// Set expiry to prevent tickets from being mined in the next
+				// sdiff interval.  When the next block begins the new interval,
+				// the ticket is being purchased for the next interval; therefore
+				// increment expiry by a full sdiff window size to prevent it
+				// being mined in the interval after the next.
+				expiry = nextIntervalStart
+				if height+1 == nextIntervalStart {
+					expiry += intervalSize
+				}
+			}
+
+			cancelCtx, cancel := context.WithCancel(ctx)
+			cancels = append(cancels, cancel)
+			go func() {
+				err := vsp.buyTicket(cancelCtx, tipHeader, expiry, balanceToMaintain)
+				if err != nil {
+					switch {
+					// silence these errors
+					case errors.Is(err, errors.InsufficientBalance):
+					case errors.Is(err, context.Canceled):
+					case errors.Is(err, context.DeadlineExceeded):
+					default:
+						log.Errorf("Ticket purchasing failed: %v", err)
+					}
+					if errors.Is(err, errors.Passphrase) {
+						fatalMu.Lock()
+						fatal = err
+						fatalMu.Unlock()
+						outerCancel()
+					}
+				}
+			}()
+		}
+	}
+}
+
+// buyTicket calculate the number of tickets before purchase
+func (vsp *VSP) buyTicket(ctx context.Context, tip *wire.BlockHeader, expiry int32, maintain int64) error {
```

Comment:
If this method is allowed to buy more than 1 ticket, the following should work. Format the comment to max 80 chars per line.

```suggestion
// buyTickets purchases one or more tickets depending on the spendable balance of the selected account and the specified minimum balance to maintain.
func (vsp *VSP) buyTickets(ctx context.Context, tip *wire.BlockHeader, expiry int32, maintain int64) error {
```